Added Docker Compose

dillonco opened the pull request:
Setup Docker so that it's an easy dev installation. Docker Compose automatically installs node and rethink, then runs the npm install and run for you. I changed the Dockerfile to be a lot slimmer and tested it locally. Also I added the basic empty tables to rethink and commited the files, I'm not sure if they will work though.

Also I cleaned up the readme a bit.

dillonco committed:
Added Docker Compose

Setup Docker so that it's an easy dev installation.

kalicki commented:
Nice, but I think in use `.env` to handle environment variables
https://docs.docker.com/compose/environment-variables/#/setting-environment-variables-with-docker-compose-run

andrecrt commented on Dockerfile:
```diff
@@ -1,30 +1,23 @@
 EXPOSE 8000
 
-CMD ["supervisord", "-c", "/etc/supervisor/supervisord.conf", "-n"]
+CMD ["npm", "run", "dev"]
```
This Dockerfile is intended to be used in any environment, not only `development`

kalicki replied:
@dillonco @andreftavares  My suggestion is to have a docker for dev and prod
dev -> `CMD ["npm", "run", "dev"]`
 prod -> `CMD ["npm", "start"]`
Ref: https://nodejs.org/en/docs/guides/nodejs-docker-webapp/

```
"scripts": {
    "dev": "NODE_ENV=development nodemon index | bunyan",
    "start": "NODE_ENV=production node index.js"
  },
```

kalicki replied:
Generally all PaaS supports NODE, execute ' start ' and he takes care of scales, another alternative is to use and pm2 not use docker

andrecrt commented on README.md:
```diff
@@ -1,17 +1,14 @@
-
-### API Resources
-After deploying and running Atlas locally, you can access the [Swagger](https://openapis.org/) API Documentation by browsing to `localhost:8000/docs`.
+![Swagger API Documentation](http://i.imgur.com/WENl84g.png "Swagger API Documentation")
```
Why can't we keep the screenshot on the repo?

dillonco replied:
I read a couple of posts like this one that advise against it, unless the pictures are part of engine: https://robinwinslow.uk/2013/06/11/dont-ever-commit-binary-files-to-git/https://robinwinslow.uk/2013/06/11/dont-ever-commit-binary-files-to-git/

andrecrt commented:
I don't think adding the database files is a good practice. Indeed there would be nice for the database to be initialised, but I think it should be via some script/fixtures that would do it programatically.

dillonco commented:
Thanks for the comments! Yeah we can store the environment variables in an .env file, I like that. I'll remove the database files and work on adding a script to populate the database automatically.

Let me make some changes to the request.

kalicki commented:
I think this for create DB and Tables:

```
let dbExists = !!(await rethinkdb.dbList()).includes(config.database.name);
            if (!dbExists) {
                rethinkdb.dbCreate(config.database.name).run();
                rethinkdb.db(config.database.name).tableCreate("Carts").run();
                rethinkdb.db(config.database.name).tableCreate("Checkouts").run();
                rethinkdb.db(config.database.name).tableCreate("Collections").run();
                rethinkdb.db(config.database.name).tableCreate("Contents").run();
                rethinkdb.db(config.database.name).tableCreate("Orders").run();
                rethinkdb.db(config.database.name).tableCreate("Products").run();
                rethinkdb.db(config.database.name).tableCreate("Users").run();
```

Not been tested but it is an idea of the app create the tables if there is

dillonco committed:
Multiple Changes

dillonco commented:
I made the following changes:
- removed database files
- added @kalicki create database script, it creates all the tables but spits out an error, next time you restart atlas it works! Probably still reporting an old error, I can fix it tomorrow.
- added .env file where users can choose their environment
- Dockerfile now looks for NODE_ENV variable for deciding what to run, defaults to dev if nothing.
- Both npm start and npm run prod do the same thing

Still some more work to be done but it's late again, let me know your comments and I'll keep improving on the pull request.

dillonco committed:
Fixed Docker Compose

dillonco commented:
Latest commit creates a database if it doesn't exist. It also allows you to set the environment through the .env file. Works pretty well, give it a try! I also want to add automatic JWT key generation to the dockerfile. I'll close this request and send a new proper one later.
